src/python/src/grpc/_adapter/_types.py

```diff
@@ -0,0 +1,365 @@
+  SSL_TARGET_NAME_OVERRIDE = 'grpc.ssl_target_name_override'
+
+@enum.unique
+class CallError(enum.IntEnum):
+  """Mirrors grpc_call_error in the C core."""
+  OK                        = 0
+  ERROR                     = 1
+  ERROR_NOT_ON_SERVER       = 2
+  ERROR_NOT_ON_CLIENT       = 3
+  ERROR_ALREADY_ACCEPTED    = 4
+  ERROR_ALREADY_INVOKED     = 5
+  ERROR_NOT_INVOKED         = 6
+  ERROR_ALREADY_FINISHED    = 7
+  ERROR_TOO_MANY_OPERATIONS = 8
+  ERROR_INVALID_FLAGS       = 9
+  ERROR_INVALID_METADATA    = 10
+
+@enum.unique
+class StatusCode(enum.IntEnum):
+  """Mirrors grpc_status_code in the C core."""
+  OK                  = 0
+  CANCELLED           = 1
+  UNKNOWN             = 2
+  INVALID_ARGUMENT    = 3
+  DEADLINE_EXCEEDED   = 4
+  NOT_FOUND           = 5
+  ALREADY_EXISTS      = 6
+  PERMISSION_DENIED   = 7
+  RESOURCE_EXHAUSTED  = 8
+  FAILED_PRECONDITION = 9
+  ABORTED             = 10
+  OUT_OF_RANGE        = 11
+  UNIMPLEMENTED       = 12
+  INTERNAL            = 13
+  UNAVAILABLE         = 14
+  DATA_LOSS           = 15
+  UNAUTHENTICATED     = 16
+
+@enum.unique
+class OpType(enum.IntEnum):
+  """Mirrors grpc_op_type in the C core."""
+  SEND_INITIAL_METADATA   = 0
+  SEND_MESSAGE            = 1
+  SEND_CLOSE_FROM_CLIENT  = 2
+  SEND_STATUS_FROM_SERVER = 3
+  RECV_INITIAL_METADATA   = 4
+  RECV_MESSAGE            = 5
+  RECV_STATUS_ON_CLIENT   = 6
+  RECV_CLOSE_ON_SERVER    = 7
+
+@enum.unique
+class EventType(enum.IntEnum):
+  """Mirrors grpc_completion_type in the C core."""
+  QUEUE_SHUTDOWN  = 0
+  QUEUE_TIMEOUT   = 1  # if seen on the Python side, something went horridly wrong
+  OP_COMPLETE     = 2
+
+class Status(collections.namedtuple(
+    'Status', [
+        'code',
+        'details',
+    ])):
+  """The end status of a GRPC call.
+
+  Attributes:
+    code (StatusCode): ...
+    details (str): ...
+  """
+
+class CallDetails(collections.namedtuple(
+    'CallDetails', [
+        'method',
+        'host',
+        'deadline',
+    ])):
+  """Provides information to the server about the client's call.
+
+  Attributes:
+    method (str): ...
+    host (str): ...
+    deadline (float): ...
+  """
+
+class OpArgs(collections.namedtuple(
+    'OpArgs', [
+        'type',
+        'initial_metadata',
+        'trailing_metadata',
+        'message',
+        'status',
+    ])):
+  """Arguments passed into a GRPC operation.
+
+  Attributes:
+    type (OpType): ...
+    initial_metadata (sequence of 2-sequence of str): Only valid if type ==
+      OpType.SEND_INITIAL_METADATA, else is None.
+    trailing_metadata (sequence of 2-sequence of str): Only valid if type ==
+      OpType.SEND_STATUS_FROM_SERVER, else is None.
+    message (bytes): Only valid if type == OpType.SEND_MESSAGE, else is None.
+    status (Status): Only valid if type == OpType.SEND_STATUS_FROM_SERVER, else
+      is None.
+  """
+
+  @staticmethod
+  def send_initial_metadata(init_md):
```

Comment:
Avoid abbreviations unless you have a really good reason. Why shouldn't this method's parameter just be "initial_metadata"?